### PR TITLE
Remove trailing slashes from source paths (#6102)

### LINKED ISSUE #6102: [Bug] Duplicated source when seed and model paths overlap
### Is this a new bug in dbt-core?

- [X] I believe this is a new bug in dbt-core
- [X] I have searched the existing issues, and I could not find an existing issue for this bug

### Current Behavior

If I have the following structure:
```
proj_name/
- data_provider
  - seeds
  - models
    - sources.yml
- < consider the previous structure repeated for N providers >
```

And the following configuration:
```yml
model-paths: 
  - proj_name/data_provider/
seed-paths: 
  - proj_name/data_provider
```

I will get an error like:
```
10:32:20  Running with dbt=1.3.0
10:32:21  Encountered an error:
Compilation Error
  dbt found two sources with the name "some_source".
  
  Since these resources have the same name, dbt will be unable to find the correct resource
  when looking for source("data_provider", "some_source").
  
  To fix this, change the name of one of these resources:
  - source.proj_name.data_provider.some_source (proj_name/data_provider/models/sources.yml)
  - source.proj_name.data_provider.some_source (proj_name/data_provider/models/sources.yml)
```

I am aware that I can do:
```yml
model-paths: 
  - proj_name/data_provider/models
seed-paths: 
  - proj_name/data_provider/seeds
```
But this is not practical for several data providers, furthermore, as the data in question is usually not mixed with other providers, we would like to keep the structure as close as it is.

### Expected Behavior

No name clash.

My intuition says that sources shouldn't be analysed when searching for seeds, but there may be a reason behind this.

### Steps To Reproduce

Described in current behavior.

### Relevant log output

```shell
Described in current behavior.
```

### Environment

```markdown
- OS: MacOS
- Python: 3.8.10
- dbt: 1.3.0
```

### Which database adapter are you using with dbt?

snowflake

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by jmg-duarte**
resolves #6102

### Description

Paths like "models" and "models/" are the same but were resolved as different. Leading to issues when seemingly equivalent (but different) paths were used in (for example) seeds and models.

The following example would detect the same files twice.

```yaml
model-paths:
  - models
seed-paths:
  - models/
```

### Checklist

- [X] I have read [the contributing guide](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md) and understand what's expected of me
- [X] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
- [X] I have run this code in development and it appears to resolve the stated issue
- [X] This PR includes tests, or tests are not required/relevant for this PR
- [X] I have [opened an issue to add/update docs](https://github.com/dbt-labs/docs.getdbt.com/issues/new/choose), or docs changes are not required/relevant for this PR
- [x] I have run `changie new` to [create a changelog entry](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md#adding-a-changelog-entry)

### CODE REVIEW

**Review by peterallenwebb — APPROVED** [review]

**Comment by peterallenwebb** [comment]
@jmg-duarte, thanks for this contribution.

It appears that to make mypy happy, which is necessary before merge, you will need to avoid assigning the paths variable twice, but with objects of different types. Overall the fix makes sense to me, and I'll keep an eye out for your update so that we can quickly merge.

**Comment by jmg-duarte** [comment]
Hey @peterallenwebb, I'm having some trouble updating the branch because I deleted the forked repo a while ago by mistake and haven't been able to push changes to the branch. I've followed guides but to no avail. I've also contacted GitHub support to see if I can avoid creating a new PR but if you feel this is important enough, I can go one to close this one and create a new branch.

**Comment by peterallenwebb** [comment]
@jmg-duarte That would be great! If don't feel like going through all the trouble, I can also get this change in on your behalf with no further effort on your part, just let me know.

**Comment by jmg-duarte** [comment]
@peterallenwebb GitHub support was super helpful and I got it working. I've applied the changes, let me know if there is something I should update further!

**Comment by jmg-duarte** [comment]
@peterallenwebb GitHub support was super helpful and I got it working. I've applied the changes, let me know if there is something I should update further!

File: core/dbt/config/project.py [hunk 1]
```diff
@@ -1,51 +1,34 @@
...
-    ProjectContractError,
-    DbtRuntimeError,
-)
+from dbt.contracts.project import PackageConfig
+from dbt.contracts.project import Project as ProjectContract
+from dbt.contracts.project import SemverString
+from dbt.dataclass_schema import ValidationError
+from dbt.exceptions import (DbtProjectError, DbtRuntimeError,
+                            ProjectContractBrokenError, ProjectContractError,
+                            SemverError)
```
Comment (emmyoop) [inline core/dbt/config/project.py:20]:
```suggestion
from dbt.exceptions import (
    DbtProjectError,
    DbtRuntimeError,
    ProjectContractBrokenError,
    ProjectContractError,
    SemverError,
)
```

File: core/dbt/config/project.py [hunk 2]
```diff
@@ -1,51 +1,34 @@
...
-
-import hashlib
-import os
+from typing import Any, Dict, List, Mapping, Optional, TypeVar, Union
 
-from dbt import flags, deprecations
-from dbt.clients.system import path_exists, resolve_path_from_base, load_file_contents
+from dbt import deprecations, flags
+from dbt.clients.system import (load_file_contents, path_exists,
+                                resolve_path_from_base)
```
Comment (emmyoop) [inline core/dbt/config/project.py:10]:
```suggestion
from dbt.clients.system import (
    load_file_contents,
    path_exists,
    resolve_path_from_base,
)
```

File: core/dbt/config/project.py [hunk 3]
```diff
@@ -1,51 +1,34 @@
...
-    selector_config_from_data,
-    selector_data_from_root,
-    SelectorConfig,
-)
+from dbt.version import get_installed_version
+from typing_extensions import Protocol, runtime_checkable
 
+from .renderer import DbtProjectYamlRenderer
+from .selectors import (SelectorConfig, selector_config_from_data,
+                        selector_data_from_root)
```
Comment (emmyoop) [inline core/dbt/config/project.py:31]:
```suggestion
from .selectors import (
    SelectorConfig,
    selector_config_from_data,
    selector_data_from_root
)
```

**Review by emmyoop — CHANGES_REQUESTED** [review]
A few formatting issues and you'll need to resolve the conflict.  Then this should be good to merge in!

**Comment by jmg-duarte** [comment]
> A few formatting issues and you'll need to resolve the conflict. Then this should be good to merge in!

Rebasing over `main` was enough, thank you! Resolved and ready for review!

File: core/dbt/config/project.py [hunk 4]
```diff
@@ -137,11 +137,10 @@ def _all_source_paths(
...
 ) -> List[str]:
-    # We need to turn a list of lists into just a list, then convert to a set to
-    # get only unique elements, then back to a list
-    return list(
-        set(list(chain(model_paths, seed_paths, snapshot_paths, analysis_paths, macro_paths)))
-    )
+    paths = chain(model_paths, seed_paths, snapshot_paths, analysis_paths, macro_paths)
+    # Strip trailing slashes since the path is the same even though the name is not
+    paths = map(lambda s: s.rstrip("/"), paths)
+    return list(set(paths))
```
Comment (emmyoop) [inline core/dbt/config/project.py:143]: my doesn't like that you change the type of `paths` here
```suggestion
    paths_stripped = map(lambda s: s.rstrip("/"), paths)
    return list(set(paths_stripped))
```

**Review by emmyoop — APPROVED** [review]
Almost there!  Just need to fix mypy.

**Review by emmyoop — CHANGES_REQUESTED** [review]
Clicked the wrong button, meant to request a change!

File: core/dbt/config/project.py (hunk 4 above)
Comment (jmg-duarte) [reply core/dbt/config/project.py:143]: I did this before, with the constant updates to the underlying code I must've messed up the rebase. Let me fix that

**Review by emmyoop — CHANGES_REQUESTED** [review]
@jmg-duarte almost there!  black now fails on a formatting issue.  If you run pre-commit (`pre-commit run` after installing it as described in the [contributing doc](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md#pre-commit)), it will make the fixes for you.  Alternatively, you can check out the failing tests and manually fix the formatting issue.

**Comment by jmg-duarte** [comment]
Should be working now. My VSCode applied isort which in turn f'd up the black formatting. Thank you for the patience

**Review by emmyoop — APPROVED** [review]
Looks great!  Thanks @jmg-duarte for the contribution.  I'll get it merged in!

**Pull request merged**